### PR TITLE
Add sanity checks to app generate upgrade plan

### PR DESCRIPTION
**Pull request by rgildein**
Like this we can create app without these error beeing raised and thay will be raised during generation of upgrade plan.

### CODE REVIEW

**Comment by rgildein** [comment]
I did not add the sanity check for `_check_mismatched_versions`, since that will be required too much work now without COUUnit.  I will add it after COUUnit (now only Unit) will be introduced, so here or in different PR.

**Review by gabrielcocenza — COMMENTED** [review]
> I did not add the sanity check for _check_mismatched_versions, since that will be required too much work now without COUUnit.

It looks like that you are adding `_check_mismatched_versions` but not the unit tests. Wouldn't worth adding it later with the unit tests once the `COUUnit` is implemented on main?

**Comment by rgildein** [comment]
> > I did not add the sanity check for _check_mismatched_versions, since that will be required too much work now without COUUnit.
> 
> It looks like that you are adding `_check_mismatched_versions` but not the unit tests. Wouldn't worth adding it later with the unit tests once the `COUUnit` is implemented on main?

I'll add it later with PR with `COUUnit` implementation, or by another PR, or if `COUUnit` land on main before this is merged, I will add here. But now I could not add `COUUnit`, so many PRs are blocking it. 
I thought that I was clear with my comment there, but it seems that not.

**Review by gabrielcocenza — APPROVED** [review]

File: cou/apps/base.py [hunk 1]
```diff
@@ -678,3 +684,65 @@ def _get_wait_step(self) -> PostUpgradeStep:
...
+
+        if self.config["enable-auto-restarts"].get("value") is False:
+            raise ApplicationError(
+                "COU does not currently support upgrading applications that disable service "
+                "restarts. Please enable charm option enable-auto-restart and rerun COU to "
+                f"upgrade the {self.name} application."
+            )
+
+    def _check_application_target(self, target: OpenStackRelease) -> None:
+        """Check if application release is not lower than or equal to target.
```
Comment (aieri) [inline cou/apps/base.py:709]: non-blocking: this is accurate, but could be more clearly expressed flipping the negation
```suggestion
        """Check if the application is already upgraded.
```

File: cou/apps/subordinate.py [hunk 2]
```diff
@@ -25,6 +26,23 @@
 class SubordinateBase(OpenStackApplication):
     """Subordinate base class."""
 
+    def _check_application_target(self, target: OpenStackRelease) -> None:
+        """Check if application release is not lower than or equal to target.
```
Comment (aieri) [inline cou/apps/subordinate.py:30]: as above

File: cou/apps/subordinate.py [hunk 3]
```diff
@@ -25,6 +26,23 @@
...
+
+        Subordinate application are not comparing apt_source_codename with target.
```
Comment (aieri) [inline cou/apps/subordinate.py:32]:
```suggestion
        Subordinate applications use the apt source of the related principal and don't have an origin/openstack-origin config option.
```

**Review by aieri — APPROVED** [review]
lgtm, I just have rewording suggestions. Feel free to move them in a separate PR if you prefer.

File: cou/apps/base.py [hunk 4]
```diff
@@ -83,14 +83,11 @@ class OpenStackApplication:
     :type units: list[ApplicationUnit]
     :raises ApplicationError: When there are no compatible OpenStack release for the
         workload version.
-    :raises MismatchedOpenStackVersions: When units part of this application are running mismatched
-        OpenStack versions.
-    :raises HaltUpgradePlanGeneration: When the class halts the upgrade plan generation.
```
Comment (agileshaw) [inline cou/apps/base.py:88]: non-blocking: why are we removing these exceptions docstrings? I'm curious about the general expectation of what and what not to include in the class docstring.

File: tests/unit/apps/test_base.py [hunk 5]
```diff
@@ -41,3 +43,93 @@ def test_application_get_latest_os_version_failed(
...
+
+
+@patch("cou.apps.base.OpenStackApplication.channel", new_callable=PropertyMock)
+@patch("cou.apps.base.OpenStackApplication._verify_channel")
+@patch("cou.apps.base.OpenStackApplication.apt_source_codename", new_callable=PropertyMock)
+@patch("cou.apps.base.OpenStackApplication.current_os_release", new_callable=PropertyMock)
+@patch(
+    "cou.apps.base.OpenStackApplication.can_upgrade_current_channel",
+    new_callable=PropertyMock(return_value=False),
+)
```
Comment (agileshaw) [inline tests/unit/apps/test_base.py:104]: non-blocking: maybe its better to include a test case with `can_upgrade_current_channel.return_value=True` (and other conditions as `false`)

**Review by agileshaw — APPROVED** [review]
LGTM with two non-blocking suggestions

**Review by gabrielcocenza — APPROVED** [review]

File: cou/apps/base.py (hunk 4 above)
Comment (rgildein) [reply cou/apps/base.py:88]: Anything, what __init__, __new__ or __post_init__ can raised.

File: tests/unit/apps/test_base.py (hunk 5 above)
Comment (rgildein) [reply tests/unit/apps/test_base.py:104]: You are right, but I will add it in another PR, since this one is blocking me.

**Pull request merged**